### PR TITLE
Replace copy coroutine implementation with pthread implementation.

### PR DESCRIPTION
**Pull request by ioquatix**
The copy coroutine is unique in that it uses a private stack area. It is also tricky to implement as it uses `setjmp`/`longjmp` to update the stack & instruction pointer. Because of this, pointers to stack allocated data are only valid while the fiber is running, and this makes implementation tricky for any code which uses stack allocated structs.

Previously we introduced some macros for dealing with this - copy coroutines fall back to `malloc`/`free` rather than using stack locals, but the chance if this causing problems is very high. In addition, `rb_protect` prevents us from switching fibers, which also prevents the fiber scheduler from operating within this context.

We propose to remove the copy coroutine implementation and replace it with a pthread-based cooperative scheduled implementation. This implementation uses one native thread per fiber, but only allows one of them to execute at a time using a mutex. This approach should be more consistent with the other native implementations and allow us to simplify a number of implementation details.

https://bugs.ruby-lang.org/issues/18015

### CODE REVIEW

**Review by jeremyevans — COMMENTED** [review]
I fully support this direction. In terms of implementation, this is quite a bit outside of my area of expertise, but I don't have any issues with it.

I'm guessing the test failures are expected.  Not sure if we want to add `argument` to the other coroutine implementations, or use `#ifdef` around the related code.  Seems like adding `argument` to other coroutine implementations would be simplest, and I don't think the additional 4/8 bytes per fiber is too high a cost.

File: coroutine/pthread/Context.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+{
+    struct coroutine_shared * shared = _shared;
+    pthread_mutex_unlock(&shared->guard);
+}
+
+void * coroutine_trampoline(void * _context)
+{
+    struct coroutine_context * context = _context;
+
+    fprintf(stderr, "coroutine_trampoline context=%p shraed=%p\n", context, context->shared);
```
[the remark this answers is not in the document]
Comment (ioquatix) [reply coroutine/pthread/Context.c:51]: Thanks, yes just debugging.

**Comment by ioquatix** [comment]
> I'm guessing the test failures are expected. Not sure if we want to add argument to the other coroutine implementations, or use #ifdef around the related code.

I think we should pass the argument into the coroutine because it isolates the design from global/per-thread state. I'll think about what makes the most sense.

**Comment by ioquatix** [comment]
Okay, this is looking pretty good, will aim to merge today. I've run about 10 full builds and have not seen any issues.

**Comment by jeremyevans** [comment]
By "full builds" does this include `make check`?  It would be fine if the pthread implementation does as well on `make check` as the copy implementation.  If not, it may still be acceptable, but it would depend on what the additional failures are.

**Comment by ioquatix** [comment]
Yes, it's now part of GH Actions and you can see everything is passing.

**Comment by ioquatix** [comment]
Okay, this is now merged.

**Pull request closed without merging**